// @4c/layout/lib/PropTypes.js
"use strict";

exports.__esModule = true;
exports.defaultProps = exports.propTypes = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var align = _propTypes["default"].oneOf(['start', 'end', 'flex-start', 'flex-end', 'center', 'stretch', 'baseline', 'first-baseline', 'last-baseline']);

var content = _propTypes["default"].oneOf(['left', 'right', 'flex-start', 'flex-end', 'center', 'baseline', 'first-baseline', 'last-baseline', 'space-between', 'space-around', 'space-evenly']);

var propTypes = {
  as: _propTypes["default"].elementType.isRequired,
  direction: _propTypes["default"].oneOf(['row', 'column']).isRequired,
  pad: _propTypes["default"].oneOfType([_propTypes["default"].bool, _propTypes["default"].number]),
  wrap: _propTypes["default"].bool,
  grow: _propTypes["default"].bool,
  inline: _propTypes["default"].bool,
  flex: _propTypes["default"].any,
  align: align.isRequired,
  alignSelf: align,
  alignContent: content,
  justify: content.isRequired
};
exports.propTypes = propTypes;
var defaultProps = {
  as: 'div',
  direction: 'row',
  align: 'stretch',
  justify: 'flex-start'
};
exports.defaultProps = defaultProps;
